storyblok/field-type-examples
# feat: use `z.preprocess` to parse the entire options object

Issue: EXT-1665

## What?

Using `z.preprocess` to parse the options object in a single call to `safeParse()`. `z.preprocess` allows us to parse the JSON string before we validate it with a schema.

One thing to note here is that `z.safeParse` is not completely safe, because if the processor in 
 `z.preprocess()` throw an exception, `safeParse` will throw that error. Therefore, we wrap `safeParse` in an actually safe function -- `superSafeParse`.

We also use `z.transform` to sort the array of languages while we are parsing the schema.

Also memorizing the options and content that we parse in `App.tsx`.

Breaking changes:

- The `highlightStates` option now require the user to also defined the default option. This will be the first element in the array. Thus, the users are required to provide at least two objects in `higlightStates` (or omit it entirely).

Outside of `parseOptions`, nothing has changed.

## Why?

We use `preprocess` and `transform` because that allows us to parse the entire object in a declarative way, without having to manually break up the parsing into different steps. 

## How to test? (optional)
 
See if the tests cover all use cases.

Manually try different combinations of options, with different values. See if the error messages are helpful.
